fifteen_min/src/viewer.rs

```diff
@@ -65,68 +58,24 @@ impl State<SimpleApp> for Viewer {
         ctx.canvas_movement();
 
         if ctx.redo_mouseover() {
-            let scale_factor = if ctx.canvas.cam_zoom >= app.opts.min_zoom_for_detail {
-                1.0
-            } else {
-                10.0
-            };
-            self.hovering_on_bldg = match app.mouseover_unzoomed_buildings(ctx) {
-                Some(ID::Building(hover_id)) => match self.hovering_on_bldg.take() {
-                    Some(previous_hover)
-                        if (previous_hover.id, previous_hover.scale_factor)
-                            == (hover_id, scale_factor) =>
-                    {
-                        Some(previous_hover)
-                    }
-                    _ => {
-                        debug!("drawing new hover");
-                        let drawn_route = self
-                            .isochrone
-                            .path_to(&app.map, hover_id)
-                            .and_then(|path| path.trace(&app.map, Distance::ZERO, None))
-                            .map(|polyline| {
-                                let dashed_lines = polyline.dashed_lines(
-                                    Distance::meters(0.75 * scale_factor),
-                                    Distance::meters(1.0 * scale_factor),
-                                    Distance::meters(0.4 * scale_factor),
-                                );
-                                let mut batch = GeomBatch::new();
-                                batch.extend(Color::BLACK, dashed_lines);
-                                ctx.prerender.upload(batch)
-                            });
-                        Some(HoverOnBuilding {
-                            id: hover_id,
-                            tooltip: if let Some(time) =
-                                self.isochrone.time_to_reach_building.get(&hover_id)
-                            {
-                                Text::from(Line(format!("{} away", time)))
-                            } else {
-                                Text::from(Line("This is more than 15 minutes away"))
-                            },
-                            scale_factor,
-                            drawn_route,
-                        })
-                    }
-                },
-                _ => None,
-            };
-
-            // Also update this to conveniently get an outline drawn
-            app.current_selection = self.hovering_on_bldg.as_ref().map(|h| ID::Building(h.id));
+            let isochrone = &self.isochrone;
+            self.hovering_on_bldg
+                .update(HoverOnBuilding::key(ctx, app), |key| {
```

Comment:
The semantics here remind me of `or_insert_with_key` for [HashMap](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with_key) and [BTreeMap](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.or_insert_with_key) - a name closer to that might be clearer - Something like:

```
self.hovering_on_bldg.get_or_insert_with_key(HoverOnBuilding::key(ctx, app), |key| {
    HoverOnBuilding::value(ctx, app, key, isochrone)
})
```

Reply:
I guess `insert` is a weird thing for a non-collection.

I don't like `update` though... 😆 

maybe `get_or_set_with_key` ?

Reply:
To expound - I don't like `update` because if we're using the cached value, nothing is `update`d.

Reply:
I guess, thinking about it a bit more, it's not a getter. I think that's what I was failing to understand. I'm much less offended by `update` now. 😬

Reply:
Right, it doesn't return anything. The naming could be `update_if_changed` or something, but I was hoping calling the type `Cached` makes this clear. Any other ideas, or is `update` OK?